src/Solvers/matrix_solver_utils.jl

```diff
@@ -95,3 +95,114 @@ end
+is provided, then the `input` and `output` fields inherit the default boundary conditions
+according to the `template_field`.
+
+For `architecture = CPU()` the matrix returned is a `SparseArrays.SparseMatrixCSC`; for `GPU()`
+is a `CUDA.CuSparseMatrixCSC`.
+"""
+function compute_matrix_for_linear_operation(::CPU, template_field, linear_operation!, args...;
+                                             boundary_conditions_input=nothing,
+                                             boundary_conditions_output=nothing)
+
+    loc = location(template_field)
+    Nx, Ny, Nz = size(template_field)
+    grid = template_field.grid
+
+    # allocate matrix A
+    A = spzeros(eltype(grid), Nx*Ny*Nz, Nx*Ny*Nz)
+
+    if boundary_conditions_input === nothing
+        boundary_conditions_input = FieldBoundaryConditions(grid, loc, template_field.indices)
+    end
+
+    if boundary_conditions_output === nothing
+        boundary_conditions_output = FieldBoundaryConditions(grid, loc, template_field.indices)
+    end
+
+    # allocate fields eᵢⱼₖ and Aeᵢⱼₖ = A*eᵢⱼₖ
+     eᵢⱼₖ = Field(loc, grid; boundary_conditions=boundary_conditions_input)
+    Aeᵢⱼₖ = Field(loc, grid; boundary_conditions=boundary_conditions_output)
+
+    for k in 1:Nz, j in 1:Ny, i in 1:Nx
+        parent(eᵢⱼₖ)  .= 0
+        parent(Aeᵢⱼₖ) .= 0
+
+        eᵢⱼₖ[i, j, k] = 1
+
+        fill_halo_regions!(eᵢⱼₖ)
+
+        linear_operation!(Aeᵢⱼₖ, eᵢⱼₖ, args...)
+
+        A[:, Ny*Nx*(k-1) + Nx*(j-1) + i] .= vec(Aeᵢⱼₖ)
+    end
+
+    return A
+end
+
+function compute_matrix_for_linear_operation(::GPU, template_field, linear_operation!, args...;
+                                             boundary_conditions_input=nothing,
+                                             boundary_conditions_output=nothing)
+
+    loc = location(template_field)
+    Nx, Ny, Nz = size(template_field)
+    grid = template_field.grid
+
+    if boundary_conditions_input === nothing
+        boundary_conditions_input = FieldBoundaryConditions(grid, loc, template_field.indices)
+    end
+
+    if boundary_conditions_output === nothing
+        boundary_conditions_output = FieldBoundaryConditions(grid, loc, template_field.indices)
+    end
+
+    # allocate fields eᵢⱼₖ and Aeᵢⱼₖ = A*eᵢⱼₖ; A is the matrix to be computed
+     eᵢⱼₖ = Field(loc, grid; boundary_conditions=boundary_conditions_input)
+    Aeᵢⱼₖ = Field(loc, grid; boundary_conditions=boundary_conditions_output)
+
+    colptr = CuArray{Int}(undef, Nx*Ny*Nz + 1)
+    rowval = CuArray{Int}(undef, 0)
+    nzval  = CuArray{eltype(grid)}(undef, 0)
+
+    CUDA.@allowscalar colptr[1] = 1
+
+    for k in 1:Nz, j in 1:Ny, i in 1:Nx
+        parent(eᵢⱼₖ)  .= 0
+        parent(Aeᵢⱼₖ) .= 0
+
+        CUDA.@allowscalar eᵢⱼₖ[i, j, k] = 1
+
+        fill_halo_regions!(eᵢⱼₖ)
+
+        linear_operation!(Aeᵢⱼₖ, eᵢⱼₖ, args...)
+
+        count = 0
+        for n in 1:Nz, m in 1:Ny, l in 1:Nx
+            Aeᵢⱼₖₗₘₙ = CUDA.@allowscalar Aeᵢⱼₖ[l, m, n]
```

Comment:
should we make a kernel?

Reply:
it's only for CPU
is there still advantage? I guess it can be speedup by the number of threads?

Reply:
Yes! Good point.

Reply:
I think I'd like to merge this so that I can move the AMGX stuff out of Oceananigans. So creation of a kernel can wait...